### PR TITLE
Fix false positive warning in `check_function_inputs_match`

### PR DESCRIPTION
**Pull request by yiyuezhuo**
# Description

* When non-function callable objects (e.x. methods and objects with `__call__` defined) is used with `EventData`, false positive warnings are raised. See #3795. It can be fixed by relaxing the function checking.
* A new `get_type_hints` is added with some tests.
* #3795 is fixed.
* No additional dependencies 

Closes: #3795 

# Checklist:

- [X] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [X] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: test/test_utils.py [hunk 1]
```diff
@@ -587,3 +590,51 @@ def test_abspath_no_symlink(self):
...
+        class C:
+            def f(self, s: str, evt: EventData):
+                return s
+
+        def f(s: str, evt: EventData):
+            return s
+
+        test_objs = [F(), C().f, f]
+
+        with warnings.catch_warnings():
```
Comment (freddyaboulton) [inline test/test_utils.py:636]: This test isn't testing anything I thin. Is the intention to check that no warnings are being emitted?

**Review by freddyaboulton — COMMENTED** [review]
Looks good @yiyuezhuo ! Thanks for the contribution. Left a comment about the test. Please add a line to the `CHANGELOG` so we can properly credit you with this change.

File: test/test_utils.py (hunk 1 above)
Comment (yiyuezhuo) [reply test/test_utils.py:636]: Yes, this test will fail if the previous code is used.

Comment (freddyaboulton) [reply test/test_utils.py:636]: Should be good to go once the CHANGELOG is fixed

**Review by freddyaboulton — APPROVED** [review]
Thanks @yiyuezhuo ! Will merge after CI passes

**Pull request merged**